src/Veldrid.Utilities/ObjParser.cs

```diff
@@ -84,140 +136,147 @@ private class ParseContext
+                        "An unsupported line-type specifier, '{0}', was used on line {1}.",
+                        piece0.ToString(),
+                        _currentLine));
                 }
             }
 
-            private void DiscoverMaterialLib(string libName)
+            private void DiscoverMaterialLib(ReadOnlySpan<char> libName)
             {
                 if (_materialLibName != null)
                 {
                     throw new ObjParseException(
-                        string.Format("mtllib appeared again in the file. It should only appear once. Line {0}, \"{1}\"", _currentLine, _currentLineText));
+                        $"mtllib appeared again in the file. It should only appear once. Line {_currentLine}.");
                 }
 
-                _materialLibName = libName;
+                _materialLibName = libName.ToString();
             }
 
-            private void ProcessFaceLine(string[] pieces)
+            private void ProcessFaceLine(
+                ref ReadOnlySpanSplitter<char> splitter,
+                ReadOnlySpan<char> piece1, ReadOnlySpan<char> piece2)
             {
-                string first = pieces[1];
-                ObjFile.FaceVertex faceVertex0 = ParseFaceVertex(first);
+                ObjFile.FaceVertex faceVertex0 = ParseFaceVertex(piece1);
 
-                for (int i = 0; i < pieces.Length - 3; i++)
+                while (splitter.MoveNext())
                 {
-                    string second = pieces[i + 2];
-                    ObjFile.FaceVertex faceVertex1 = ParseFaceVertex(second);
-                    string third = pieces[i + 3];
-                    ObjFile.FaceVertex faceVertex2 = ParseFaceVertex(third);
+                    ReadOnlySpan<char> piece3 = splitter.Current;
+                    ObjFile.FaceVertex faceVertex1 = ParseFaceVertex(piece2);
+                    ObjFile.FaceVertex faceVertex2 = ParseFaceVertex(piece3);
 
                     DiscoverFace(new ObjFile.Face(faceVertex0, faceVertex1, faceVertex2, _currentSmoothingGroup));
+                    piece2 = piece3;
                 }
             }
 
-            private ObjFile.FaceVertex ParseFaceVertex(string faceComponents)
+            private ObjFile.FaceVertex ParseFaceVertex(ReadOnlySpan<char> faceComponents)
             {
-                string[] slashSplit = faceComponents.Split(s_slashChar, StringSplitOptions.None);
-                if (slashSplit.Length != 1 && slashSplit.Length != 2 && slashSplit.Length != 3)
-                {
-                    throw CreateExceptionForWrongFaceCount(slashSplit.Length);
-                }
+                if (faceComponents.IsEmpty)
+                    throw CreateExceptionForWrongFaceCount();
 
-                int pos = ParseInt(slashSplit[0], "the first face position index");
+                int firstSlash = faceComponents.IndexOf(_slashChar);
+                ReadOnlySpan<char> firstPart = faceComponents.Slice(firstSlash + 1);
```

Comment:
IMO, the naming here is a little confusing, because "firstPart" actually refers to the second part of the line in my thinking. Then, "firstSlice" is defined below, but that one makes sense for its name. This is effectively "afterFirstSlash" -- it's the Span for everything after the first slash.

Reply:
I'll try to clarify